Signal/src/ViewControllers/ContactViewController.swift

```diff
@@ -0,0 +1,686 @@
+        case .unknown:
+            let activityIndicator = UIActivityIndicatorView(activityIndicatorStyle: .whiteLarge)
+            topView.addSubview(activityIndicator)
+            activityIndicator.autoPinEdge(.top, to: .bottom, of: lastView, withOffset: 10)
+            activityIndicator.autoHCenterInSuperview()
+            lastView = activityIndicator
+            break
+        }
+
+        lastView.autoPinEdge(toSuperviewEdge: .bottom, withInset: 15)
+
+        let bottomView = UIView.container()
+        bottomView.backgroundColor = UIColor.white
+        bottomView.layoutMargins = .zero
+        bottomView.preservesSuperviewLayoutMargins = false
+        rootView.addSubview(bottomView)
+        bottomView.autoPinEdge(.top, to: .bottom, of: topView, withOffset: 0)
+        bottomView.autoPinEdge(toSuperviewEdge: .bottom)
+        bottomView.autoPinEdge(.left, to: .left, of: self.view, withOffset: 0)
+        bottomView.autoPinEdge(.right, to: .right, of: self.view, withOffset: 0)
+        bottomView.setContentHuggingVerticalLow()
+
+        var lastRow: UIView?
+
+        let addSpacerRow = {
+            guard let prevRow = lastRow else {
+                owsFail("\(self.logTag) missing last row")
+                return
+            }
+            let row = UIView()
+            row.backgroundColor = UIColor(rgbHex: 0xdedee1)
+            bottomView.addSubview(row)
+            row.autoSetDimension(.height, toSize: 1)
+            row.autoPinLeadingToSuperviewMargin(withInset: self.hMargin)
+            row.autoPinTrailingToSuperviewMargin()
+            row.autoPinEdge(.top, to: .bottom, of: prevRow, withOffset: 0)
+            lastRow = row
+        }
+
+        let addRow: ((UIView) -> Void) = { (row) in
+            if lastRow != nil {
+                addSpacerRow()
+            }
+            bottomView.addSubview(row)
+            row.autoPinLeadingToSuperviewMargin()
+            row.autoPinTrailingToSuperviewMargin()
+            if let lastRow = lastRow {
+                row.autoPinEdge(.top, to: .bottom, of: lastRow, withOffset: 0)
+            } else {
+                row.autoPinEdge(toSuperviewEdge: .top, withInset: 0)
+            }
+            lastRow = row
+        }
+
+        if viewMode == .nonSystemContact {
+            addRow(createActionRow(labelText: NSLocalizedString("CONVERSATION_SETTINGS_NEW_CONTACT",
+                                                               comment: "Label for 'new contact' button in conversation settings view."),
+                                   action: #selector(didPressCreateNewContact)))
+
+            addRow(createActionRow(labelText: NSLocalizedString("CONVERSATION_SETTINGS_ADD_TO_EXISTING_CONTACT",
+                                                               comment: "Label for 'new contact' button in conversation settings view."),
+                                   action: #selector(didPressAddToExistingContact)))
+        }
+
+        // TODO: Not designed yet.
```

Comment:
There's a bunch of TODOs here that correspond to cards or items in the "open design qs" doc.